### PR TITLE
[WIP] Vyper support

### PR DESCRIPTION
**Pull request by leonprou**
This is a WIP PR of the Vyper support.

### CODE REVIEW

**Comment by leonprou** [comment]
That's the WIP PR for vyper: https://github.com/leonprou/vyper/tree/add_output_ast_support

Instructions to make it running
```
download my vyper repo and install dependencies.
export VYPER=path_to_cloned_vyper_repo
cd to this repo and switch to dev mode
PATH=$PATH:$VYPER/bin PYTHONPATH=$PYTHONPATH:$VYPER slither ./examples/vyper/crownfund.vy
```

File: slither/slither_vyper.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        '''
+            Args:
+                contract (str)
+            Keyword Args:
+                vyper (str): Vyper binary location (default 'vyper')
+        '''
+        self._detectors = []
+        self._printers = []
+        self._triage_mode = False
+        super(Slither, self).__init__(filename)
```
Comment (disconnect3d) [inline slither/slither_vyper.py]:
```suggestion
        super().__init__(filename)
```
In Python 3, you just call  `super()` and the  magic happens under the hood

File: slither/slither_vyper.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+        f = open(contract)
+        kode = f.read()
+        code = parse_to_ast(kode)
+        global_ctx = GlobalContext.get_global_context(code)
+        self._global_ctx = global_ctx
+
+    def _run_vyper(self, filename, vyper):
+        if not os.path.isfile(filename):
+            logger.error('{} does not exist (are you in the correct directory?)'.format(filename))
```
Comment (disconnect3d) [inline slither/slither_vyper.py]:
```suggestion
            logger.error('%s does not exist (are you in the correct directory?)', filename)
```

File: slither/slither_vyper.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        else:
+            cmd = [vyper, '-f', 'ast', filename]
+
+            process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+            stdout, stderr = process.communicate()
+            stdout, stderr = stdout.decode(), stderr.decode()  # convert bytestrings to unicode strings
+
+            if stderr:
+                stderr = stderr.split('\n')
```
Comment (disconnect3d) [inline slither/slither_vyper.py]:
```suggestion
                stderr = stderr.splitlines()
```

File: slither/slither_vyper.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                stderr = [x if 'Error' not in x else red(x) for x in stderr]
```
Comment (disconnect3d) [inline slither/slither_vyper.py]:
```suggestion
                stderr = (x if 'Error' not in x else red(x) for x in stderr)
```

File: slither/vyper_parsing/declarations/contract.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+from slither.slithir.variables import StateIRVariable
+
+from slither.vyper_parsing.variables.state_variable import StateVariableVyper
+
+
+logger = logging.getLogger("ContractVyperParsing")
+
+class ContractVyper(Contract):
+    def __init__(self, slitherVyper, data):
+        super(ContractVyper, self).__init__()
```
Comment (disconnect3d) [inline slither/vyper_parsing/declarations/contract.py]:
```suggestion
        super().__init__()
```

File: slither/vyper_parsing/declarations/function.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import logging
+
+from slither.core.declarations.function import Function
+
+logger = logging.getLogger("FunctionVyper")
+
+class FunctionVyper(Function):
+    def __init__(self, function, function_sig, contract):
+        super(FunctionVyper, self).__init__()
```
Comment (disconnect3d) [inline slither/vyper_parsing/declarations/function.py]:
```suggestion
        super().__init__()
```

File: slither/vyper_parsing/declarations/structure.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+class StructureVyper(Structure):
+    """
+    Structure class
+    """
+    # elems = [(type, name)]
+
+
+    def __init__(self, name, ast_struct, code):
+        super(StructureVyper, self).__init__()
```
Comment (disconnect3d) [inline slither/vyper_parsing/declarations/structure.py]:
```suggestion
        super().__init__()
```

File: slither/vyper_parsing/slither_vyper.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import json
+
+from slither.core.slither_core import Slither
+from slither.vyper_parsing.declarations.contract import ContractVyper
+
+logger = logging.getLogger("VyperParsing")
+
+class SlitherVyper(Slither):
+    def __init__(self, filename):
+        super(SlitherVyper, self).__init__()
```
Comment (disconnect3d) [inline slither/vyper_parsing/slither_vyper.py]:
```suggestion
        super().__init__()
```

File: slither/vyper_parsing/variables/event_variable.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        """
+        Analyze event variable attributes
+        :param attributes: The event variable attributes to parse.
+        :return: None
+        """
+
+        # Check for the indexed attribute
+        self._indexed = indexed
+
+        super(EventVariableVyper, self)._analyze_variable_attributes(arg_sig)
```
Comment (disconnect3d) [inline slither/vyper_parsing/variables/event_variable.py]:
```suggestion
        super()._analyze_variable_attributes(arg_sig)
```

File: slither/vyper_parsing/variables/variable_declaration.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+    def __init__(self, var_record):
+        '''
+            A variable can be declared through a statement, or directly.
+            If it is through a statement, the following children may contain
+            the init value.
+            It may be possible that the variable is declared through a statement,
+            but the init value is declared at the VariableDeclaration children level
+        '''
+        super(VariableDeclarationVyper, self).__init__()
```
Comment (disconnect3d) [inline slither/vyper_parsing/variables/variable_declaration.py]:
```suggestion
        super().__init__()
```

**Comment by montyly** [comment]
Good work!

We should remove the `vyper` import, and make the ast output from vyper an util (ex: `vyper-ast-json` in https://github.com/ethereum/vyper/tree/master/bin). I understand that it makes the dev of the support faster, but I think its better to not keep the vyper's dependency (to avoid issues on vyper's codebase and update).

**Comment by leonprou** [comment]
@montyly thanks!

Yeah, I get your concerns about the `vyper` dep, might be problematic. `vyper-ast-json` may be a good idea, do you in touch with the vyper team?

So do you prefer first to work on cutting the `vyper imports`, or continue using them and work on adding new functionality and cut these imports latter?

**Comment by jacqueswww** [comment]
Wow this cool :) For the Vyper dependency you might consider an option of being able to install vyper as an optional dependency, as in `pip install slither[vyper]`. And then you could lazy import/wrap the required vyper imports. This also makes it clear which vyper version you are supporting.

As for the serializer/deserializer `vyper-ast-json` that'd be great to be part of Vyper itself, and is something we want as part of our AST Classes.

If you have any question feel free to pop over to our gitter.

**Comment by montyly** [comment]
@leonprou yes I think cutting the `vyper imports` will prevent architectural issues later.

We can go for a naive version of `vyper-ast-json`, and iterate over it. We are likely to have missing features, that will need to be added. 

It would definitely make a lot of sense if we can contribute to https://github.com/ethereum/vyper/issues/1363. Though I am ok if, first, we have a simple version based on the current codebase, and we synchronize with the new vyper parser following it's evolution.

File: slither/slither_vyper.py (hunk 2 above)
Comment (leonprou) [reply slither/slither_vyper.py]: I actually prefer to use the `format` syntax. Any objections for it?

**Comment by leonprou** [comment]
@disconnect3d went over your code comments, fixed all but one. Anyway I copied most of the stuff from the slither solc code, so this got to be fixed there also :grimacing:

**Comment by montyly** [comment]
Hi @leonprou, are you still working on the issue?
Do you need any help from my side?

**Comment by leonprou** [comment]
@montyly Hey, been busy lately so can't find time working on this issue.
Also got stuck on the parsing function content, which is pretty complex.

So I think it would make sense to move the bounty to someone else. I'll try to submit some more work I've done till now.

**Comment by jacqueswww** [comment]
Just a BTW: Not sure if this helps, but vyper now has JSON style output using the `-f ast` flag ;)

**Comment by montyly** [comment]
Wow great @jacqueswww, thanks for letting us know!

**Comment by montyly** [comment]
I am closing this PR in favor of https://github.com/crytic/slither/pull/300

**Pull request closed without merging**